core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp

```diff
@@ -72,17 +90,12 @@ class Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>,
         *space.impl_internal_space_instance();
     sycl::queue& q = *instance.m_queue;
 
-    auto parallel_for_event = q.submit([functor, policy](sycl::handler& cgh) {
+    auto parallel_for_event = q.submit([&](sycl::handler& cgh) {
+      FunctorWrapperRangePolicyParallelFor<Functor, Policy> f{policy.begin(),
+                                                              functor};
       sycl::range<1> range(policy.end() - policy.begin());
-      const auto begin = policy.begin();
-
-      cgh.parallel_for(range, [=](sycl::item<1> item) {
-        const typename Policy::index_type id = item.get_linear_id() + begin;
-        if constexpr (std::is_same<WorkTag, void>::value)
-          functor(id);
-        else
-          functor(WorkTag(), id);
-      });
+      cgh.parallel_for<FunctorWrapperRangePolicyParallelFor<Functor, Policy>>(
```

Comment:
The crucial change is being able to provide the template argument, `FunctorWrapperRangePolicyParallelFor<Functor, Policy>`, here (and I couldn't get that to work with the lambda). 
Using a functor also matches the change for https://github.com/kokkos/kokkos/pull/4097.